[WebGPU] api,validation,queue,* is failing

mwyrzykowski opened the pull request:
#### 65fb3ec0feaee6c99d6dbcf47914b54c69f640ae
<pre>
[WebGPU] api,validation,queue,* is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=267279">https://bugs.webkit.org/show_bug.cgi?id=267279</a>
&lt;radar://120725451&gt;

Reviewed by Tadeu Zagallo.

Add validation for GPUQueue and related operations along with passing exepctations.

* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/buffer_mapped-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/buffer_mapped.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/buffer-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/buffer.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/texture-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/texture.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/submit.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeBuffer.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeTexture-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeTexture.spec.js:
Add passing expectations and update tests to latest revision of the CTS.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::isOriginClean):
(WebCore::dimension):
(WebCore::isStateValid):
(WebCore::convertRGBA8888ToRGB10A2):
(WebCore::copyToDestinationFormat):
(WebCore::GPUQueue::copyExternalImageToTexture):
Fix reporting errors when a cross origin asset is used.

* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
(WGSL::AttributeValidator::parseLocation):

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):

* Source/WebGPU/WebGPU/BindableResource.h:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::setCommandEncoder const):
(WebGPU::Buffer::destroy):
(WebGPU::Buffer::mapAsync):
(WebGPU::Buffer::unmap):
Add validation for buffer mapping.

* Source/WebGPU/WebGPU/CommandBuffer.h:
* Source/WebGPU/WebGPU/CommandBuffer.mm:
(WebGPU::CommandBuffer::setBufferMapCount):
(WebGPU::CommandBuffer::bufferMapCount const):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::errorValidatingCopyBufferToBuffer):
(WebGPU::CommandEncoder::incrementBufferMapCount):
(WebGPU::CommandEncoder::decrementBufferMapCount):
(WebGPU::CommandEncoder::copyBufferToBuffer):
(WebGPU::validateImageCopyBuffer):
(WebGPU::validateCopyBufferToTexture):
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::makeSubmitInvalid):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::CommandEncoder::finish):
(WebGPU::CommandEncoder::resolveQuerySet):
(WebGPU::CommandEncoder::writeTimestamp):
(WebGPU::CommandEncoder::validateCopyBufferToBuffer): Deleted.
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::setCommandEncoder):
(WebGPU::ComputePassEncoder::setBindGroup):
* Source/WebGPU/WebGPU/ExternalTexture.h:
* Source/WebGPU/WebGPU/QuerySet.h:
(WebGPU::QuerySet::isValid const): Deleted.
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::QuerySet::isValid const):
(WebGPU::QuerySet::destroy):
(WebGPU::QuerySet::setCommandEncoder const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateSubmit const):
(WebGPU::Queue::submit):
(WebGPU::Queue::validateWriteBuffer const):
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(WebGPU::validateWriteBufferInitial): Deleted.
* Source/WebGPU/WebGPU/RenderBundle.h:
* Source/WebGPU/WebGPU/RenderBundle.mm:
(-[ResourceUsageAndRenderStage initWithUsage:renderStages:entryUsage:binding:resource:]):
(-[ResourceUsageAndRenderStage initWithUsage:renderStages:entryUsage:binding:]): Deleted.
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):
(WebGPU::RenderBundleEncoder::setBindGroup):
(WebGPU::RenderBundleEncoder::setIndexBuffer):
(WebGPU::RenderBundleEncoder::setVertexBuffer):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::drawIndexedIndirect):
(WebGPU::RenderPassEncoder::drawIndirect):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::setCommandEncoder):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::parentEncoder):
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setIndexBuffer):
(WebGPU::RenderPassEncoder::setVertexBuffer):
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::setCommandEncoder const):
(WebGPU::Texture::destroy):
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::destroy):
(WebGPU::TextureView::setCommandEncoder const):
(WebGPU::TextureView::setCommandEncoder): Deleted.

Canonical link: <a href="https://commits.webkit.org/273436@main">https://commits.webkit.org/273436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/448af86b2764cf122eb3a0c909d881704e8132d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35368 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38105 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31894 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36564 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11337 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30754 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31492 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10596 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10635 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31608 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39351 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32132 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31940 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36611 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10795 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8697 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34645 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12550 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8100 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11319 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/html/OffscreenCanvas.h:
```diff
@@ -158,6 +158,7 @@ class OffscreenCanvas final : public ActiveDOMObject, public RefCounted<Offscree
     void dispatchEvent(Event&) final;
     using RefCounted::ref;
     using RefCounted::deref;
+    bool isDetached() const { return m_detached; };
```
@cdumez @mattwoodrow this is the only non-WebGPU change. We need to know if an OffscreenCanvas is detached to pass some of the tests.

I'm not aware of an alternative method of determining that other than adding a new member function on OffscreenCanvas to return that internal state variable.

djg commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -328,16 +329,182 @@ static void imageBytesForSource(const auto& source, const auto& destination, Ima
+#if PLATFORM(COCOA)
+static uint32_t convertRGBA8888ToRGB10A2(uint8_t r, uint8_t g, uint8_t b, uint8_t a)
 {
+    uint32_t r0 = static_cast<uint32_t>(r) * 4 + 1;
```
Is this right? Eg, 255 will be converted to 1021, which is just a little bit smaller than 1023, such that 255/255 == 1.0f != 1021/1023.

I would expect that a correct conversion would keep the same float representation (as near as possible).

mwyrzykowski replied:
Oh indeed. I’ll change this to:

`(c * 1023) / 255`

instead, thanks for catching

mwyrzykowski replied:
thanks for catching @djg I fixed the alpha calculation which was really wrong as well

djg replied:
If you want to avoid the `mul` and `div` you might be able to get away with copying the 2 msbs from the source into the 2 lsbs of the dest. ie:

```
r0 = ((uint32_t) r << 2) | ((uint32_t) r >> 6)
```

djg replied:
My ex-boss, Charles, it probably the authority on the issues involved. http://cbloomrants.blogspot.com/2021/04/requantization-to-different-unorm-bit.html

mwyrzykowski replied:
I'll take that change :) The values are effectively the same and in some cases the bit shifting might be more accurate as it emulates rounding - https://onlinegdb.com/dKdKd1gzo

mwyrzykowski committed:
[WebGPU] api,validation,queue,* is failing
https://bugs.webkit.org/show_bug.cgi?id=267279
<radar://120725451>

Reviewed by Tadeu Zagallo.

Add validation for GPUQueue and related operations along with passing exepctations.

* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/buffer_mapped-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/buffer_mapped.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/buffer-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/buffer.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/texture-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/destroyed/texture.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/submit.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeBuffer.spec.js:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeTexture-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/queue/writeTexture.spec.js:
Add passing expectations and update tests to latest revision of the CTS.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::isOriginClean):
(WebCore::dimension):
(WebCore::isStateValid):
(WebCore::convertRGBA8888ToRGB10A2):
(WebCore::copyToDestinationFormat):
(WebCore::GPUQueue::copyExternalImageToTexture):
Fix reporting errors when a cross origin asset is used.

* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
(WGSL::AttributeValidator::parseLocation):

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):

* Source/WebGPU/WebGPU/BindableResource.h:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::setCommandEncoder const):
(WebGPU::Buffer::destroy):
(WebGPU::Buffer::mapAsync):
(WebGPU::Buffer::unmap):
Add validation for buffer mapping.

* Source/WebGPU/WebGPU/CommandBuffer.h:
* Source/WebGPU/WebGPU/CommandBuffer.mm:
(WebGPU::CommandBuffer::setBufferMapCount):
(WebGPU::CommandBuffer::bufferMapCount const):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::errorValidatingCopyBufferToBuffer):
(WebGPU::CommandEncoder::incrementBufferMapCount):
(WebGPU::CommandEncoder::decrementBufferMapCount):
(WebGPU::CommandEncoder::copyBufferToBuffer):
(WebGPU::validateImageCopyBuffer):
(WebGPU::validateCopyBufferToTexture):
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::makeSubmitInvalid):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::CommandEncoder::finish):
(WebGPU::CommandEncoder::resolveQuerySet):
(WebGPU::CommandEncoder::writeTimestamp):
(WebGPU::CommandEncoder::validateCopyBufferToBuffer): Deleted.
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::setCommandEncoder):
(WebGPU::ComputePassEncoder::setBindGroup):
* Source/WebGPU/WebGPU/ExternalTexture.h:
* Source/WebGPU/WebGPU/QuerySet.h:
(WebGPU::QuerySet::isValid const): Deleted.
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::QuerySet::isValid const):
(WebGPU::QuerySet::destroy):
(WebGPU::QuerySet::setCommandEncoder const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateSubmit const):
(WebGPU::Queue::submit):
(WebGPU::Queue::validateWriteBuffer const):
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(WebGPU::validateWriteBufferInitial): Deleted.
* Source/WebGPU/WebGPU/RenderBundle.h:
* Source/WebGPU/WebGPU/RenderBundle.mm:
(-[ResourceUsageAndRenderStage initWithUsage:renderStages:entryUsage:binding:resource:]):
(-[ResourceUsageAndRenderStage initWithUsage:renderStages:entryUsage:binding:]): Deleted.
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):
(WebGPU::RenderBundleEncoder::setBindGroup):
(WebGPU::RenderBundleEncoder::setIndexBuffer):
(WebGPU::RenderBundleEncoder::setVertexBuffer):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::drawIndexedIndirect):
(WebGPU::RenderPassEncoder::drawIndirect):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::setCommandEncoder):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::parentEncoder):
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setIndexBuffer):
(WebGPU::RenderPassEncoder::setVertexBuffer):
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::setCommandEncoder const):
(WebGPU::Texture::destroy):
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::destroy):
(WebGPU::TextureView::setCommandEncoder const):
(WebGPU::TextureView::setCommandEncoder): Deleted.

Canonical link: https://commits.webkit.org/273436@main
